vpr/src/pack/prepack.cpp

```diff
@@ -742,65 +817,73 @@ t_pack_molecule *alloc_and_load_pack_molecules(
-	 * Simplifying assumptions: Each atom can map to at most one molecule,
-     *                          use first-fit mapping based on priority of pattern
-	 * TODO: Need to investigate better mapping strategies than first-fit
-     */
-	for (i = 0; i < num_packing_patterns; i++) {
-		best_pattern = 0;
-		for(j = 1; j < num_packing_patterns; j++) {
-			if(is_used[best_pattern]) {
-				best_pattern = j;
-			} else if (is_used[j] == false && compare_pack_pattern(&list_of_pack_patterns[j], &list_of_pack_patterns[best_pattern]) == 1) {
-				best_pattern = j;
-			}
-		}
-		VTR_ASSERT(is_used[best_pattern] == false);
-		is_used[best_pattern] = true;
+	std::vector<bool> feasible_patterns(num_packing_patterns);
 
         auto blocks = atom_ctx.nlist.blocks();
         for(auto blk_iter = blocks.begin(); blk_iter != blocks.end(); ++blk_iter) {
             auto blk_id = *blk_iter;
 
-            cur_molecule = try_create_molecule(list_of_pack_patterns, atom_molecules, best_pattern, blk_id);
-            if (cur_molecule != nullptr) {
-                cur_molecule->next = list_of_molecules_head;
-                /* In the event of multiple molecules with the same atom block pattern,
-                 * bias to use the molecule with less costly physical resources first */
-                /* TODO: Need to normalize magical number 100 */
-                cur_molecule->base_gain = cur_molecule->num_blocks - (cur_molecule->pack_pattern->base_cost / 100);
-                list_of_molecules_head = cur_molecule;
-
-                //Note: atom_molecules is an (ordered) multimap so the last molecule
-                //      inserted for a given blk_id will be the last valid element
-                //      in the equal_range
-                auto rng = atom_molecules.equal_range(blk_id); //The range of molecules matching this block
-                bool range_empty = (rng.first == rng.second);
-                bool cur_was_last_inserted = false;
-                if (!range_empty) {
-                    auto last_valid_iter = --rng.second; //Iterator to last element (only valid if range is not empty)
-                    cur_was_last_inserted = (last_valid_iter->second == cur_molecule);
-                }
-                if(range_empty || !cur_was_last_inserted) {
-                    /* molecule did not cover current atom (possibly because molecule created is
-                     * part of a long chain that extends past multiple logic blocks), try again */
-                    --blk_iter;
-                }
-            }
+		/* Find forced pack patterns
+		 * Simplifying assumptions: Each atom can map to at most one molecule,
+		 */
+		for (int pattern = 0; pattern < num_packing_patterns; pattern++) {
+			feasible_patterns[pattern] = check_atom_feasible_for_pattern(blk_id, &list_of_pack_patterns[pattern]);
+		}
+		best_pattern = find_best_pattern(blk_id, list_of_pack_patterns, feasible_patterns, num_packing_patterns, atom_molecules);
```

Comment:
There is a possible bug I believe.  This code consumes the entire chain from the block it starts at.  How are we sure that the starting block is the root of the entire chain?  Do we need logic to walk down the chain?

Reply:
Block in the netlist are ordered and get them here always starting from the root. Maybe this assumption is too far fetched here. It may be safer to walk down the chain, but I'm not sure if necessary.